src/conways.rs

```diff
@@ -1,25 +1,263 @@
-#[derive(Clone, PartialEq)]
+use std::usize;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
 pub enum CellState {
     Dead,
     Alive,
 }
 
+#[derive(Debug)]
 pub struct Grid {
     grid: Vec<Vec<CellState>>,
+    width: usize,
+    height: usize,
 }
 
 impl Grid {
     pub fn new(width: usize, height: usize) -> Self {
         Self {
             grid: vec![vec![CellState::Dead; width]; height],
+            width,
+            height,
         }
     }
 
-    pub fn set(&mut self, x: usize, y: usize) {
+    pub fn set_alive(&mut self, x: usize, y: usize) {
         self.grid[y][x] = CellState::Alive;
     }
 
     pub fn get(&self, x: usize, y: usize) -> CellState {
         self.grid[y][x].clone()
     }
+
+    /// Calculate the next generation of cells based on the rules of Conway's Game of life
+    /// https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life
+    pub fn next_cell_generation(&mut self) {
+        let mut new_grid = vec![vec![CellState::Dead; self.width]; self.height];
+
+        for y in 0..self.height {
+            for x in 0..self.width {
+                let neighbors = self.count_neighbors(x, y);
+                let current_state = &self.grid[y][x];
+
+                new_grid[y][x] = match (current_state, neighbors) {
+                    // Rule 1: Any live cell with fewer than two live neighbors dies
+                    (CellState::Alive, 0..=1) => CellState::Dead,
+                    // Rule 2: Any live cell with two or three live neighbors lives
+                    (CellState::Alive, 2..=3) => CellState::Alive,
+                    // Rule 3: Any live cell with more than three live neighbors dies
+                    (CellState::Alive, 4..=8) => CellState::Dead,
+                    // Rule 4: Any dead cell with exactly three live neighbors becomes alive
+                    (CellState::Dead, 3) => CellState::Alive,
+                    // All other cells remain in their current state
+                    (state, _) => state.clone(),
+                };
+            }
+        }
+
+        self.grid = new_grid;
+    }
+
+    /// Count the number of alive neighbors for a given cells
+    fn count_neighbors(&self, x: usize, y: usize) -> u8 {
+        let mut count = 0;
+
+        for dy in -1..=1 {
+            for dx in -1..=1 {
+                // Skip the center cell (the cell itself)
+                if dx == 0 && dy == 0 {
+                    continue;
+                }
+
+                let nx = x as i32 + dx;
+                let ny = y as i32 + dy;
+
+                // Only count if within grid boundaries
+                if (nx >= 0 && nx < self.width as i32 && ny >= 0 && ny < self.height as i32)
+                    && self.grid[ny as usize][nx as usize] == CellState::Alive
```

Comment:
Here you can use `get(idx)` and work with other `Option` functions like `and_then`, `is_some`, etc. With that, you don't need to check if the `idx` is outside the vector.

Reply:
Done!